### PR TITLE
Better handling of invalid or duplicate content lengths.

### PR DESCRIPTION
**Pull request by alyssawilk**
And with tests for zero byte bodies.  Sorry I missed that corner case in the first pass.

### CODE REVIEW

File: source/common/http/http1/codec_impl.cc [hunk 1]
```diff
@@ -411,8 +411,9 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
-    } else if (parser_.method == http_method::HTTP_POST ||
-               parser_.method == http_method::HTTP_PUT) {
+    } else if ((parser_.method == http_method::HTTP_POST ||
+                parser_.method == http_method::HTTP_PUT) &&
+               headers->ContentLength() == nullptr) {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: nit: I think the following comment is a bit misleading. It should be something like: "if content-length is not specified for certain methods..." Can we also reference new RFC section which says that we should do this? The rules have changed slightly at least according to new RFC and I'm not even completely sure we should do this change. (It's probably OK for HTTP/1.1, but will it break anyone else?) cc @jpinner-lyft

Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: Does envoy support half-close semantics?  If a POST/PUT is not chunked and there's no content length specified I'm not sure how else end-of-body would be communicated.

Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: On server-side we don't support half closed. On client side we will correctly process response is client closes. Do you know what happens with current code for HTTP/1.1 if you POST without content-length? I'm actually not sure. Might be good to run your test without the fix to see what happens.

Ultimately in thinking about I agree this change makes sense, I would just update the comments a bit if you don't mind. (I'm just petrified of changes like this and I should have been more careful when the first one went by. We always wind up breaking some badly written app when we do something like this).

Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: I hear you.   If the risks to prod traffic are not worth it on your end I can back this part out.    It reminds me though - it's worth talking at some point about strategies we've used internally to mitigate high risk changes to not lose binary pushes, and see if any are worth incorporating into Envoy :-)

File: source/common/http/http1/codec_impl.cc [hunk 2]
```diff
@@ -411,6 +411,13 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
+    } else if ((parser_.method == http_method::HTTP_POST ||
+                parser_.method == http_method::HTTP_PUT) &&
+               headers->ContentLength() == nullptr) {
+      // If content length is not specified for certain methods and the request
+      // is not chunked encodided, return an error to the user.
```
Comment (dnoe) [inline source/common/http/http1/codec_impl.cc]: nit, typo here.

File: source/common/http/http1/codec_impl.cc (hunk 1 above)
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: Yup happy to chat. In general I would put a change like this behind "runtime" control or doing a stat first, but not sure if it's worth doing all that extra work if we can feel comfortable with it.

However, I'm still honestly not sure this is correct or we might need to expand. The reason is that AFAIK in the new RFC, content-length is totally decoupled from method. So, I think for HTTP/1.1, we need to rely on caller actually ending the HTTP/1.1 message and potentially not do this check? See comment here: https://github.com/lyft/envoy/blob/master/source/common/http/http1/codec_impl.cc#L398

File: source/common/http/http1/codec_impl.cc [hunk 3]
```diff
@@ -471,11 +478,13 @@ void ServerConnectionImpl::onResetStream(StreamResetReason reason) {
 
 void ServerConnectionImpl::sendProtocolError() {
   // We do this here because we may get a protocol error before we have a logical stream. Higher
-  // layers can only operate on streams, so there is no coherent way to allow them to send a 400
+  // layers can only operate on streams, so there is no coherent way to allow them to send an error
   // "out of band." On one hand this is kind of a hack but on the other hand it normalizes HTTP/1.1
   // to look more like HTTP/2 to higher layers.
   if (!active_request_ || !active_request_->response_encoder_.startedResponse()) {
-    Buffer::OwnedImpl bad_request_response("HTTP/1.1 400 Bad Request\r\n"
+    Buffer::OwnedImpl bad_request_response("HTTP/1.1 " + std::to_string(enumToInt(error_code_)) +
```
Comment (dnoe) [inline source/common/http/http1/codec_impl.cc]: Building this response string could be done with fmt::format(), might be clearer (and maybe faster?)

File: source/common/http/http1/codec_impl.cc (hunk 1 above)
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: So, the change for example might be to add content-length: 0 for higher layers if it's missing, on message complete? Just thinking out loud now.

Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: I guess this comes back to my question before on what http-parser does if content-length is omitted and there is no chunk encoding. IIRC it infers content length is 0 during parsing.

Comment (jpinner-lyft) [inline source/common/http/http1/codec_impl.cc]: RFC 7230 Sec. 3.3: https://tools.ietf.org/html/rfc7230#section-3.3
```
   The presence of a message body in a request is signaled by a
   Content-Length or Transfer-Encoding header field.  Request message
   framing is independent of method semantics, even if the method does
   not define any use for a message body.
```

Comment (jpinner-lyft) [inline source/common/http/http1/codec_impl.cc]: RFC 7231 Sec. 6.5.10: https://tools.ietf.org/html/rfc7231#section-6.5.10
```
   The 411 (Length Required) status code indicates that the server
   refuses to accept the request without a defined Content-Length
   (Section 3.3.2 of [RFC7230]).  The client MAY repeat the request if
   it adds a valid Content-Length header field containing the length of
   the message body in the request message.
```

Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: Yeah was going to link that. Also https://tools.ietf.org/html/rfc7230#section-3.3.3: 

"6.  If this is a request message and none of the above are true, then
       the message body length is zero (no message body is present)."

IMO as long as http-parser correctly infers content-length: 0 if none is set for request, we should probably not do this check. We should possible insert content-length: 0 for higher layers in this case?

Comment (jpinner-lyft) [inline source/common/http/http1/codec_impl.cc]: It used to be in RFC 2616 that the message body for requests dependent upon the method being sent. The bis standards loosened the requirements so that the framing layer could be decoupled from the method semantics.

Requiring a Content-Length header for POST / PUT for HTTP/1.1 should be fine; however, it may break some hand-written clients that had not been setting content-length and letting the server assume an empty entity body.

IIRC there is some security considerations to requiring Content-Length as a mitigation for HTTP Request Smuggling attacks. This assumes that the server is also correctly failing requests with multiple content-length headers that do not match.

Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: "IIRC there is some security considerations to requiring Content-Length as a mitigation for HTTP Request Smuggling attacks."
This.  But as we're not using it in production yet we can do it the slow and save way of adding stats and making sure it won't do harm.  Or alternately making sure we add an explicit content-length:0 so all layers agree.

Either way sounds like whatever we do should be done in a follow-up so I'll remove it here :-)

Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: Agreed. Let's figure this out in dedicated follow up change. Totally happy to do the right thing here but sounds like we might want to have runtime/stat control over this change. We can discuss in the context of the follow up @alyssawilk

File: source/common/http/http1/codec_impl.cc (hunk 3 above)
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: oh awesome.  I'd been trying to find the envoy equivalent of strcat and failing.  Fixed!

**Comment by mattklein123** [comment]
@alyssawilk do you mind updating the PR title? I think this is now now actually doing content length validation :)

**Comment by htuch** [comment]
@alyssawilk This is failing `check_format` in CI. Can you `fix_format` and update the PR? Ta.

**Comment by mattklein123** [comment]
nit: PR title also has a typo in it. Please fix that also.
